src/engine/LazyGroupBy.cpp

```diff
@@ -0,0 +1,310 @@
+  evaluationContext._beginIndex = beginIndex;
+  evaluationContext._endIndex = endIndex;
+
+  for (auto& aggregateAlias : aggregateAliases_) {
+    for (auto& aggregate : aggregateAlias.aggregateInfo_) {
+      // Evaluate child expression on block
+      auto exprChildren = aggregate.expr_->children();
+      sparqlExpression::ExpressionResult expressionResult =
+          exprChildren[0]->evaluate(&evaluationContext);
+
+      std::visit(
+          [this, blockSize, &evaluationContext,
+           &aggregate]<sparqlExpression::SingleExpressionResult T>(
+              T&& singleResult) {
+            auto generator = sparqlExpression::detail::makeGenerator(
+                std::forward<T>(singleResult), blockSize, &evaluationContext);
+
+            for (const auto& val : generator) {
+              addToAggregateFunction(aggregate.aggregateDataIndex_, val,
+                                     &evaluationContext);
+            }
+          },
+          std::move(expressionResult));
+    }
+  }
+}
+
+// _____________________________________________________________________________
+
+template <size_t NUM_AGGREGATED_COLS>
+void LazyGroupBy<NUM_AGGREGATED_COLS>::populateGroupBlock(
+    const IdTable& idTable, size_t row) {
+  if (currentGroupBlock_.empty()) {
+    for (size_t col : groupByCols_) {
+      currentGroupBlock_.emplace_back(col, idTable(0, col));
+    }
+  } else {
+    for (auto& columnPair : currentGroupBlock_) {
+      columnPair.second = idTable(row, columnPair.first);
+    }
+  }
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_AGGREGATED_COLS>
+bool LazyGroupBy<NUM_AGGREGATED_COLS>::rowMatchesCurrentBlock(
+    const IdTable& idTable, size_t row) {
+  return std::all_of(currentGroupBlock_.begin(), currentGroupBlock_.end(),
+                     [&](const auto& columns) {
+                       return idTable(row, columns.first) == columns.second;
+                     });
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_AGGREGATED_COLS>
+void LazyGroupBy<NUM_AGGREGATED_COLS>::substituteGroupVariable(
+    const std::vector<GroupBy::ParentAndChildIndex>& occurrences,
+    ValueId value) {
+  for (const auto& occurrence : occurrences) {
+    auto newExpression =
+        std::make_unique<sparqlExpression::IdExpression>(std::move(value));
+
+    occurrence.parent_->replaceChild(occurrence.nThChild_,
+                                     std::move(newExpression));
+  }
+}
+
+// _____________________________________________________________________________
+
+template class LazyGroupBy<0>;
+template class LazyGroupBy<1>;
+template class LazyGroupBy<2>;
+template class LazyGroupBy<3>;
+template class LazyGroupBy<4>;
+template class LazyGroupBy<5>;
+static_assert(
+    5 == DEFAULT_MAX_NUM_COLUMNS_STATIC_ID_TABLE,
+    "Please update the explicit template instantiations in `LazyGroupBy.cpp`.");
```

Comment:
As I said, we should probably do something column-based,. that doesn't need those instantiations for several reasons.